### DISCUSSION TITLE
Interface for passing a community detection algorithm to another algorithm

### DISCUSSION THREAD

**Discussion by michitux** [networkit/networkit#785] [Q&A]
I'm currently working on EgoSplitting, i.e., #482. The basic idea of the EgoSplitting algorithm is that it constructs a graph from the neighborhood of each node and applies a community detection algorithm to each of them. Further, it applies a community detection algorithm once to the graph with expanded egos. Ideally, the user should be able to freely select and configure each of these community detection algorithms. Unfortunately, this is not possible with the current `CommunityDetectionAlgorithm` interface as the graph is set in the constructor. I see basically two solutions, each with significant drawbacks which is why I would like to ask you for additional ideas or feedback which of these implementations would be acceptable to you.

* In #482, we introduce factories. There is a factory subclass for each community detection algorithm that takes the same parameters as the community detection algorithm excluding the graph. In this implementation, these factories implement a method that returns a `std::function<Partition(const Graph &)>` and the `EgoSplitting` algorithm only takes this `std::function` as parameter. This makes it easy to alternatively use a Python lambda as clustering algorithm. Alternatively, the method of the factory could also take a graph as input and return a `CommunityDetectionAlgorithm` - then EgoSplitting would take the factory as argument. The disadvantage of this approach is that we need a new factory class for every community detection algorithm that duplicates its parameters. Additionally, while the `std::function`-approach gives us full Python interoperability, not using the `Algorithm` interface feels completely wrong. Note that these are actually two options, the one implemented in #482 and the one where the factory returns a `CommunityDetectionAlgorithm`.
* As an alternative, I have started implementing an approach where a `CommunityDetectionAlgorithm` instance can be copied and re-used with a `setGraph()`-method that allows setting the actually used graph later. Then, `EgoSplitting` would simply take two `CommunityDetectionAlgorithm`-instances as parameters. Note that having one instance of each algorithm is not enough for our implementation as we need one instance per thread to be able to call them for several (small) graphs in parallel. Originally, this seemed like a great idea, but it turns out that copying objects for which you just have a base class pointer is quite ugly. Basically, you need to define a `clone()` method in the base class and then implement that `clone()` method for every derived class. This implementation can be kind of automated with an intermediate class using CRTP. After solving these initial problems, I noticed, unfortunately, that there are additional problems:
  * Some algorithms cannot be copied as they have atomics as members, e.g., `LouvainMapEquation` from #753 
  * Some algorithms, again, `LouvainMapEquation` is an example, already resize certain members in their constructor for the given graph, meaning that changing the graph later is a bad idea.

My feeling is that the first option is completely against NetworKit's interfaces which is why I started looking into the second option. However, the second option means that we need to restrict what algorithms may do in their constructor.

Do you have any other ideas, or opinions on these two options?

**Comment by avdgrinten** [comment]
What about using the second option, but instead of copying, we pass a unique_ptr?

**Reply by michitux** [reply]
The copy is required for two reasons:

* We need to run the community detection algorithm on several graphs. Some algorithms do not allow executing `run()` several times. We can (and currently do) reset `hasRun()` in `setGraph()`, but this means that we need to ensure that algorithms properly re-initialize their data structures when `run()` is called again. By copying the algorithm before executing the `run()`-method, we only need to ensure that the constructor does not already initialize data structures depending on the graph.
* We execute the community detection algorithm on many small graphs, which is why we want to do this in parallel for several graphs, i.e., run the same community detection algorithm on graph A in Thread 1, on graph B in Thread 2 etc. This is only possible if we have a separate copy of the algorithm for reach thread (or a separate copy per graph).

The different instances of the algorithm are actually stored in `unique_ptr`-objects. You can see our current implementation of the second option [in this commit](https://github.com/michitux/networkit/commit/b381839f8bd7aac458bbe8c26831f0ff4c23fb13). Note that we have since fixed `LouvainMapEquation` to be cloneable, too.

**Reply by avdgrinten** [reply]
Regarding properly re-initializing data structures in `run()`: that's something we should support anyway. I think we even have algorithms that support a `setGraph` method. (Maybe we should demand this from all new algorithms?)

In general, I think that NetworKit `Algoritihm`s should be considered non-movable (and also non-copyable), since moving around derived classes easily runs into problems and we sometimes also use non-movable members. Thus, `Algorithm` objects should always be passed around as `unique_ptr`s.

In your use case, we really do need a factory. But instead of having a `std::function` that runs the algorithm, I'd suggest taking a `std::function` that constructs the `Algorithm` (or `CommunityDetectionAlgorithm` in this case; like you describe in your first case in the OP) object but returns a unique_ptr: `std::function<std::unique_ptr<CommunityDetectionAlgorithm>()>`. We do not need a factory *class*, such a factory *function* should be enough, right?

**Reply by michitux** [reply]
I thought the change of the `Graph` references to pointers in all algorithms was explicitly to allow copying and moving algorithms, see e.g. #460? If this is not desired, wouldn't it be better to explicitly delete the corresponding methods of the `Algorithm` class? Or have I misunderstood the original idea of making algorithms assignable (which imho implies making them copyable or movable)?

Yes, a function should be enough, but the function could also be implemented as a class with a call operator that stores the parameters for the algorithm. Instead of a class, we could also have a function that creates the `std::function` with a lambda that stores the parameters. All of this would still require a function or a class that creates the `std::function`. We could implement this function that creates the `std::function` as a static function of the community detection algorithm classes such that the knowledge of the names and types of the parameters is collected in a single class (though then in two places in the class). With your example, we would still need `setGraph()`. Alternatively, the graph could also be passed as (only) parameter to the `std::function` such that the algorithm is directly constructed with the correct graph instance set.

I'm currently slightly in favor of the solution with the clone method as it seems cleaner from a code duplication (parameter names and types) point of view. But I think this is really the question if algorithms should be copyable/movable or not.
